edb/pgsql/metaschema.py

```diff
@@ -4551,36 +4554,29 @@ def _generate_schema_ver_views(schema: s_schema.Schema) -> List[dbops.View]:
 
 def _make_json_caster(
     schema: s_schema.Schema,
-    json_casts: Mapping[s_types.Type, s_casts.Cast],
     stype: s_types.Type,
     context: str,
-) -> Callable[[Any], str]:
-    cast = json_casts.get(stype)
-
-    if cast is None:
-        raise RuntimeError(
-            f'there is no direct cast from std::json to '
-            f'the type of {context!r} '
-            f'({stype.get_displayname(schema)})'
-        )
-
-    if cast.get_from_cast(schema):
-        pgtype = types.pg_type_from_object(schema, stype)
+) -> Callable[[str], str]:
+    cast_expr = qlast.TypeCast(
+        expr=qlast.TypeCast(
+            expr=qlast.Parameter(name="__replaceme__", optional=False),
+            type=s_utils.typeref_to_ast(schema, schema.get('std::json')),
+        ),
+        type=s_utils.typeref_to_ast(schema, stype),
+    )
 
-        def _cast(val: Any) -> str:
-            return f'({val})::{q(*pgtype)}'
-    else:
-        if cast.get_code(schema):
-            cast_name = cast.get_name(schema)
-            func_name = common.get_cast_backend_name(
-                cast_name, aspect='function')
-        else:
-            func_name = cast.get_from_function(schema)
+    cast_ir = qlcompiler.compile_ast_fragment_to_ir(
+        cast_expr,
+        schema,
+    )
 
-        def _cast(val: Any) -> str:
-            return f'{q(*func_name)}({val})'
+    cast_sql, _ = compiler.compile_ir_to_sql(
+        cast_ir,
+        use_named_params=True,
+        singleton_mode=True,
+    )
 
-    return _cast
+    return lambda val: cast_sql.replace('__replaceme__', val)
```

Comment:
ouch, is this really necessary?

Reply:
This is the simplest and reliable way.  SQL AST is immutable, so it isn't possible to do this easily with an AST visitor.